registry/storage/manifeststore.go

```diff
@@ -154,6 +154,49 @@ func (ms *manifestStore) Put(ctx context.Context, manifest distribution.Manifest
 // Delete removes the revision of the specified manifest.
 func (ms *manifestStore) Delete(ctx context.Context, dgst digest.Digest) error {
 	dcontext.GetLogger(ms.ctx).Debug("(*manifestStore).Delete")
+
+	if !ms.blobStore.deleteEnabled {
+		return distribution.ErrUnsupported
+	}
+
+	// Ensure the blob is available for deletion
+	_, err := ms.blobStore.blobAccessController.Stat(ctx, dgst)
+	if err != nil {
+		return err
+	}
+
+	man, err := ms.Get(ctx, dgst)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve manifest %w", err)
+	}
+
+	var subject *distribution.Descriptor
+	mt, content, _ := man.Payload()
+	switch mt {
+	case v1.MediaTypeArtifactManifest:
+		m := &ociartifact.DeserializedManifest{}
+		if err := m.UnmarshalJSON(content); err != nil {
+			return fmt.Errorf("unable to unmarshall manifest %w", err)
+		}
+		subject = m.Subject
+	case v1.MediaTypeImageManifest:
+		m := &ocischema.DeserializedManifest{}
+		if err := m.UnmarshalJSON(content); err != nil {
+			return fmt.Errorf("unable to unmarshall manifest %w", err)
+		}
+		subject = m.Subject
+	}
```

Comment:
We don't need to unmarshal the manifest again as it has been unmarhsalled by `ms.Get()`.

Reply:
Refactored this part with switch and type casting.